api/v1alpha2/component_type.go

```diff
@@ -16,3 +21,63 @@ const (
 func (ct ComponentType) GetResName(instance string) string {
 	return fmt.Sprintf("%s-%s", instance, ct)
 }
+
+func (ct ComponentType) GetObjectMeta(hdb *HStreamDB, meta *metav1.ObjectMeta) metav1.ObjectMeta {
+	fixedMeta := metav1.ObjectMeta{
+		Name:      ct.GetResName(hdb.Name),
+		Namespace: hdb.Namespace,
+		Labels: map[string]string{
+			InstanceKey:  hdb.Name,
+			ComponentKey: string(ct),
+		},
+	}
+
+	if meta != nil {
+		fixedMeta.Labels = mergeMap(fixedMeta.Labels, meta.Labels)
+		fixedMeta.Annotations = mergeMap(fixedMeta.Annotations, meta.Annotations)
+	}
+
+	return fixedMeta
+}
+
+func (ct ComponentType) GetService(hdb *HStreamDB, ports []corev1.ServicePort, isInternal bool) corev1.Service {
+	meta := ct.GetObjectMeta(hdb, nil)
+
+	if isInternal {
+		meta.Name = ct.GetResName(hdb.Name + "-internal")
+	}
+
+	return corev1.Service{
+		ObjectMeta: meta,
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{
+				ComponentKey: string(ct),
+			},
+			Ports: ports,
+		},
+	}
+}
+
+func (ct ComponentType) GetHeadlessService(hdb *HStreamDB, ports []corev1.ServicePort) corev1.Service {
+	svc := ct.GetService(hdb, ports, true)
+	svc.Spec.ClusterIP = corev1.ClusterIPNone
+	svc.Spec.PublishNotReadyAddresses = true
+
+	return svc
+}
```

Comment:
I think this should in `controller`

Reply:
My initial thought was all component types' related methods are too spread out. Converting them as struct methods is more direct to use.

It's ok for me to put them into the `controller` package. How about defining a new struct to include them as methods?